### ISSUE TITLE
jSS7 Simulator: additional MAP operations for Location Services Management (LSM)

### ISSUE THREAD

**Issue by FerUy** [RestComm/jss7#100]
Add MAP SRIforLCS, MAP PSL and MAP SLR to jSS7 simulator. 
For more information, see https://github.com/RestComm/gmlc/issues/10

**Comment by deruelle** [comment]
@f-alonso-vendrell just assigned this issue to you ;)

**Comment by f-alonso-vendrell** [comment]
Adding support for SRIforLCS in SS7 Simulator, need to do:
   * Creation of 2 new "Testing Task", MAP_LOCATION_TEST_CLIENT, MAP_LOCATION_TEST_SERVER (visual part)
  * Add the 3 operations (each request / response), e.g. taking as reference: SendRoutingInfoForLCSRequest (integration with existing implementation)
  * Integration with simulator stack to actually send/receive messages.

**Comment by FerUy** [comment]
Nice @f-alonso-vendrell 

Maybe we could put LCS instead of LOCATION (i.e. MAP_LCS_TEST_CLIENT/SERVER). What do you think? I leave it up to you.

**Comment by f-alonso-vendrell** [comment]
sure, I was trying to set up a clear todo list for my team and myself so if some concept is wrong or the approach is not ok, it can be exposed before advancing in the wrong direction. BTW how is documentation generated? I am currenly reading directly the ascii-doc but if it already generated or I can generate with a mvn command it would be of great help for my eyes.

**Comment by f-alonso-vendrell** [comment]
just discovered that at: jss7\docs\userguide\sources-asciidoc\target\generated-docs\html-book the documentation is generated.

**Comment by FerUy** [comment]
Great @f-alonso-vendrell

**Comment by FerUy** [comment]
One question, are you able to access https://github.com/RestComm/gmlc/issues/10 ?
I'm asking as they are tightly coupled.

**Comment by f-alonso-vendrell** [comment]
yes. I am reading information from there to get a picture.

**Comment by FerUy** [comment]
Great!

**Comment by vetss** [comment]
Hello @FerUy @f-alonso-vendrell 

yes we need two "Testing tasks", one is GMLC server (I guess "Client") side and another - HLR / MSC / SGSN sides ("server").

For tasks names - please check the list of MAP operations there: https://github.com/RestComm/jss7/blob/master/map/MAPMessagesImplemented.ods
I think better to give a name of a service - "Location Service Management Services" == "LSM"

MAP_LSM_TEST_CLIENT
MAP_LSM_TEST_SERVER

Please also check, which messages which side sends (GMLC and HLR/MSC/SGSN sides I mean)

Fernando, do you know which is a proper messageflow ? Is is just sending a request / responsing or the messageflow may be more complicated ?

**Comment by FerUy** [comment]
Hi @vetss 

The message flow is portrayed here: https://github.com/RestComm/gmlc/issues/10 (same diagram here)

Regarding the acronym, LCS is used by 3GPP/LTE and other related organizations for referring to Location Services. For example, 3GPP TS 23.271 is titled "Functional stage 2 description of Location Services (LCS)". 

"Location Service Management Services" is used in 3GPP TS 29.002 (MAP spec) as the section title for these location services operations:
MAP-SEND-ROUTING-INFO-FOR-LCS (MAP SRIforLCS)
MAP-PROVIDE-SUBSCRIBER-LOCATION (MAP PSL)

Another acronym largely used and known in the industry is LBS (Location Based Services), which is more commercial than technical. 

So, although I'd rather stay with LCS, what about?
MAP_LCSM_TEST_CLIENT
MAP_LCSM_TEST_SERVER


![restcomm_gmlc umts](https://cloud.githubusercontent.com/assets/14883482/15789458/b065615c-29a2-11e6-8b4e-37e169757989.png)

**Comment by vetss** [comment]
Thanks @FerUy for your info,

There is one more operation - "subscriberLocationReport - 86" that is sent TO GMLC (not from as for two others). I am not sure we need to have it. Just for your info.

As for a name of the service - go ahead in this topic and select that you like more.

**Comment by FerUy** [comment]
Hi @vetss , We will need "subscriberLocationReport - 86" for some use cases involving deferred type of location information (Notification type of Geolocation in Restcomm Geolocation API -geofencing-), but for now let's focus on SRIforLCS and PSL (Immediate type of Geolocation in Restcomm Geolocation API).

I choose LCS for the name of the service then :)
So, @f-alonso-vendrell , unless you have any objection, let's go with:
MAP_LCS_TEST_CLIENT
MAP_LCS_TEST_SERVER

**Comment by f-alonso-vendrell** [comment]
Ok, I am going with the MAP_LCS_TEST_CLIENT, created following:

 new file:   core/src/main/java/org/mobicents/protocols/ss7/tools/simulator/tests/map/TestMapLcsClientMan.java
 new file:   core/src/main/java/org/mobicents/protocols/ss7/tools/simulator/tests/map/TestMapLcsClientManMBean.java
 new file:   gui/src/main/java/org/mobicents/protocols/ss7/tools/simulatorgui/tests/map/TestMapLcsClientForm.java

and modified corresponding 

modified:   core/src/main/java/org/mobicents/protocols/ss7/tools/simulator/management/Instance_TestTask.java
modified:   core/src/main/java/org/mobicents/protocols/ss7/tools/simulator/management/TesterHost.java
modified:   gui/src/main/java/org/mobicents/protocols/ss7/tools/simulatorgui/ConnectionForm.java
modified:   gui/src/main/java/org/mobicents/protocols/ss7/tools/simulatorgui/SimulatorGuiForm.java

As the moment of this comment I have been able to show "sendRoutingInfoForLCSRequest not implemented yet" when clicking the button.

As I plan to: 
1)  hardcode operation parameters on SRIforLCS, test end2end (so I would need also SERVER Part), then pull request
2) Make parameters on SRIforLCS configurable, test end2end, then pull request
3) Move forward with the other operations

Let me know at what level of viability you would like to do pull request (e.g. is it ok to publish the SRIforLCS with parameters hardcoded or should they be configurable before publishing)

After some testing I would like to request a feature on the Simulator, that would be a "Load Default Values for Side A/B" at top level, so once you have selected Layer 1, Layer 2, Layer 3 and TestTask you click that button and all layers get populated at once.

**Comment by FerUy** [comment]
@f-alonso-vendrell agreed with the plan. It's OK to go first with 1 (publish the SRIforLCS with hardcoded parameters ) and make pull request, then go to 2 and so on.

**Comment by f-alonso-vendrell** [comment]
Hello I just got to a point in which I run 2 simulators, the first one with MAP_LCS_TEST_CLIENT and the second one with SMS_TEST_SERVER (as I haven't done the server part yet). I am able to send SRIforLCS request (at least I think so). This is what I am sending:

98114 [AWT-EventQueue-0] DEBUG org.mobicents.protocols.ss7.tools.simulator.tests.map.TestMapLcsClientMan  - MAPDialogLsm Crea
ted
98129 [AWT-EventQueue-0] DEBUG org.mobicents.protocols.ss7.tools.simulator.tests.map.TestMapLcsClientMan  - Added SendRouting
InfoForLCSRequest
98145 [AWT-EventQueue-0] DEBUG org.mobicents.protocols.ss7.sccp.impl.SccpRoutingControl  - Tx : SCCP Message=Sccp Msg [Type=-
1 networkId=0 sls=1 incomingOpc=-1 incomingDpc=-1 outgoingDpc=-1 CallingAddress(pc=1,ssn=8,AI=67,gt=null) CalledParty(pc=2,ss
n=8,AI=67,gt=null) DataLen=68]
98145 [Thread-3] DEBUG org.mobicents.protocols.sctp.AssociationImpl  - Tx : Ass=Ass_main PayloadData [dataLength=124, complet
e=true, unordered=false, payloadProtocolId=3, streamNumber=0, data=
Start: 0 (0x00)  End: 123 (0x7B)  Length: 124 (0x7C)
00: 01 00 01 01 00 00 00 7C  02 00 00 08 00 00 00 66 | .......| .......f
10: 00 06 00 08 00 00 00 65  02 10 00 64 00 00 00 01 | .......e ...d....
20: 00 00 00 02 03 02 00 01  09 01 03 07 0B 04 43 02 | ........ ......C.
30: 00 08 04 43 01 00 08 44  62 42 48 04 00 00 00 01 | ...C...D bBH.....
40: 6B 1E 28 1C 06 07 00 11  86 05 01 01 01 A0 11 60 | k.(..... .......`
50: 0F 80 02 07 80 A1 09 06  07 04 00 00 01 00 25 03 | ........ ......%.
60: 6C 1A A1 18 02 01 00 02  01 55 30 10 80 05 91 11 | l....... .U0.....
70: 11 22 22 A1 07 80 05 55  55 45 44 44 -- -- -- -- | .""....U UEDD
]

On the server part (not implemented yet LCS) I get a message which says:

00:03:15,726 WARN  [MAPStackImpl-Simulator] ApplicationContextName of non activated MAPService is received. Will send back TCAP Abort : 0, 4, 0, 0, 1, 0, 37, 3,
80550 [pool-10-thread-1] WARN org.mobicents.protocols.ss7.map.MAPStackImpl-Simulator  - ApplicationContextName of non activated MAPService is received. Will send back TCAP Abort : 0, 4, 0, 0, 1, 0, 37, 3,

which I guess makes sense since MAPServiceLsm is not active on that side. 

I would like to do a PR so you can check the code, how should I proceed?

**Comment by vetss** [comment]
> ApplicationContextName of non activated MAPService is received

check an example in org.mobicents.protocols.ss7.tools.simulator.tests.sms.TestSmsClientMan
line: 547

**Comment by f-alonso-vendrell** [comment]
Hello Vetss, so should I also do the counter part (MAP_LCS_TEST_SERVER) before PR? the client part which you are referring is working fine (I think so), the error is coming in the server part for which I haven't done anything. I will edit my previous post so that is clear.

**Comment by vetss** [comment]
Hello @FerUy 

what do you mean "counter part" ?

Yes, you need to provide both CLIENT and SERVER parts. Else it is impossible to check how well they are implemented. If both Simulator parts work fine then we can start to test of interconnecting of Simulator with some server.

**Comment by f-alonso-vendrell** [comment]
ok, proceeding...

**Comment by f-alonso-vendrell** [comment]
Server part completed, thanks @vetss for pointing out code for reference. I have pushed it into my fork. Ready to do PR.

**Comment by vetss** [comment]
Hello @f-alonso-vendrell ,

I have just checked your pull request, thanks for your work.
I can not commit your work sofar because it is not completed. My remarks are here:

1) let's rename your new package 
org.mobicents.protocols.ss7.tools.simulator.tests.map ->
org.mobicents.protocols.ss7.tools.simulator.tests.lcs
We need it because for MAP stack we have several tasks and we make names depending of test task

2) no configurable parameters. At least we need to for a client size configurable mlcNumber and targetMS. We need GUI form for updating and parameters files like TestAtiClientConfigurationData. I do not know if we need any configurable parameters at server side (ask Fernando). Possibly we need to have an opportunity to send MAP error messages But even if not we need an empty form anyway. Also we can make that server part will auto response to incoming messages (without pressing of a button in the form).

3) Response is not implemented. At least I get TC-ABORT message.

4) no information is not displayed in task forms. Please see how it is implemented in other tasks. We need at least log sent / received messages including error messages and also some reports of if a dialog is opened or not, how many messages were sent / received, etc.

5) we need to provide also access via JMX and HTTP. For this we need to have TestAtiClientStandardManMBean and register beans. Check how it is made in other tests.

6) we do not add code into master branch more because weh have just released it. We use now netty-2 branch, please take it into account. We will rename soon netty-2 to master, but now please synch with netty-2 branch

**Comment by f-alonso-vendrell** [comment]
Hello @vetss 
1) ok
2) I thought that first PR would be with hardcoded params, ok, I guess this was the next step
3) I also thought that PR would be by message, so no need to the response, as item 2, I guess this was also the next step so ok
4) Yes, I forgot to review that part. I will put messages
5) ok, wasn't aware, but I guess that is the gui connecting to a core part
6) ok

   The order of resolution will be: 1(Easy), 5(good to learn), 3(good to learn), 4, 2, 6 Let me know if once done 1,5,and 3 I can make a new PR or should I complete the whole set.

**Comment by vetss** [comment]
Hello @FerUy 

> 2) I thought that first PR would be with hardcoded params, ok, I guess this was the next step

If we are saying of work steps we can do it step by step for sure. But the final version must contain parameters editing forms / storing of parameters into some configs. Even we have as for now an empty set of configurable parameters those forms must be present so all implementation follow the same standard and furter code update will be simple

> 3) I also thought that PR would be by message, so no need to the response, 
> as item 2, I guess this was also the next step so ok

I do not understand your fully. We need to have both client a nd server parts workable and test interconnections between two simulators (client - server) to be sure all is properly implemented before we will start of testing of interconnection of SS7 Simulator with some server like GMLC or other SS7 equipment.

> 5) ok, wasn't aware, but I guess that is the gui connecting to a core part

We need to test JMX / HTTP accesses before we can say that the implementation is finished. This is of cause possible when both client and server parts are implemented. If you need some extra info for this part feel free to ask it.

**Comment by f-alonso-vendrell** [comment]
simulator is not working for me on netty-2 branch, added a comment on: https://gitter.im/RestComm/jss7

**Comment by vetss** [comment]
Hello @f-alonso-vendrell 

I do not understnad why you have such problemms in pushing your code into netty-2 branch

I think the proper way for you is:
- kust download code from netty-2 branch
- incorporate (cherry-pick or may be just copy your new files) your updates to netty-2 branch
- debug it

**Comment by mprojo** [comment]
Hi @vetss, @FerUy 

I will be working with @f-alonso-vendrell on MAP_SUBSCRIBER_LOCATION_REPORT request and response.

Could you please guide us on which parameters will require configuration?

For the request, we've initially just considered the parameters used in your map-impl test classes (see below). But not sure if this covers all you need for issue #10?

Thanks in advance!

LCSClientID lcsClientID = mapParameterFactory.createLCSClientID(LCSClientType.plmnOperatorServices, null, null,
                    null, null, null, null);
            ISDNAddressString networkNodeNumber = mapParameterFactory.createISDNAddressString(
                    this.testerHost.getConfigurationData().getTestMapLcsClientConfigurationData().getAddressNature(),
                    this.testerHost.getConfigurationData().getTestMapLcsClientConfigurationData().getNumberingPlanType(), address);
            LCSLocationInfo lcsLocationInfo = mapParameterFactory.createLCSLocationInfo(networkNodeNumber, null, null, false,
                    null, null, null, null, null);
clientDialogLsm.addSubscriberLocationReportRequest(LCSEvent.emergencyCallOrigination, lcsClientID, lcsLocationInfo,
                    null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, false, false,
                    null, null, null, null, false, null, null, null);

**Comment by vetss** [comment]
Hello @mprojo 

this operation is a rich one and a list of a needed parameters is a difficult question. We understand that implementing of all possible parameters is a hard task then we can start with a minimum set That you have described.

@FerUy if you have a concrete list of parameters that is needed as a first step, please provide it.

**Comment by vetss** [comment]
Hello @f-alonso-vendrell 

I have committed your logic from https://github.com/RestComm/jss7/pull/143 pull request
into my repo branch that is on the top of current master branch. 
https://github.com/vetss/jss7/tree/ss7-26

But I need to recheck code and then I will merge it into a master branch. Please wait some time, I will try to finish this work next week.

**Comment by FerUy** [comment]
Hi @mprojo , sorry for late response, been traveling and working overseas the last few days (and will for the next week). 

You are covering some of the mandatory parameters, which is fine. Eventually we will need quite a bunch of them as defined in clause 13A.2 in 3GPP TS 29.002 (MAP), but for our goal we are especially concerned about those involving event-based location info (or deferred MT-LR procedure). Please consider **adding** the following:

MSISDN or IMSI (or both)
LMSI
IMEI
LCS-Reference Number
Area Event Info
H-GMLC address
Age of Location Estimate
Cell Id or SAI

For these parameters, please refer to the latest version of the aforementioned specs, i.e. http://www.3gpp.org/DynaReport/29002.htm and http://www.3gpp.org/dynareport/23271.htm.

**Comment by mprojo** [comment]
Thanks for your reply @FerUy and @vetss .

Clause 13A.2 applies to MAP-PROVIDE-SUBSCRIBER-LOCATION, I understand that you need the same params for MAP-SUBSCRIBER-LOCATION-REPORT request (clause 13A.3) which I will work on, thus:
MSISDN or IMSI (or both)
LMSI
IMEI
LCS-Reference Number
H-GMLC address
Age of Location Estimate
Cell Id or SAI
In addition to mandatory params: LCS Event, LCS Client ID, Network Node Number. Just considering mandatory attributes for each param (eg LCSClientType for LCSClientID)).

For the MAP-SUBSCRIBER-LOCATION-REPORT response, we initially have been considered NA ESRD Address, but I guess we need to consider LCS-Reference Number as well to support deferred MT-LR procedure.

Thanks again!

**Comment by FerUy** [comment]
Great, thank you @mprojo !
Looking forward to seeing this beauty in action :)

**Comment by vetss** [comment]
Hello @f-alonso-vendrell 

I have checked your update in https://github.com/RestComm/jss7/pull/143 and committed it into master branch (this is the latest branch as for now).

Your provided update is not completed and I have not added my updates for it because this code needs much extra work.
My comments:
- TestMapLcsClient/ServerManMBean does not contain methods like "void putNumberingPlan(String val);" They are all for updating of enum based configurable properties via HTTP interface for values that are based on EnumeratedBase abstract class. Check how it is implemented in other examples.
- TestMapLcsClientManMBean (the same is for Server part) does not need two methods like performSendRoutingInfoForLCSRequest() and do not need lower level methods like sendRoutingInfoForLCSRequest() (the last one is needed to be private I guess)
- performSubscriberLocationReportRequest() must be implemented at the sever side from my point of view. The idea is that Client plays a role of GMLC and a Server plays a role of HLR, MSC, VLR etc.
performSendRoutingInfoForLCSRequest() is initiated by GMLC side (Client side) and performSubscriberLocationReportRequest() is initiated by VLR / MSC side (Server side). Or we can make vice versa - Server==GMLC, Clent==HLR, MSC, VLR.
- GUI does not looks like proper implemented. We have at Client side "SRI Request" and "SRC Request" that shares same parameters editing that is not corrent. At the Server side I can not see any proper implemented parameters.
- TestMapLcsClientStandardManMBean and TestMapLcsServerStandardManMBean are not fully implemented also. These classes are needed for HTTP access and must cover all attribute changes and "perform" actions.
- no response for a second implemented operation

Please synch to a master branch and try to fix missed parts.

**Comment by f-alonso-vendrell** [comment]
Hello @vetss . Ok I will sync with master branch and will do remaining implementations needed.

**Comment by vetss** [comment]
Thaks @f-alonso-vendrell

**Comment by f-alonso-vendrell** [comment]
Hello @vetss, Synced with master and put together changes done by @mprojo in my fork, some of your comments are partially covered with her changes, although since still remain some I will no be issuing new PR until I return from vacation.

BTW, latest change to 1.7 compile.source makes 4 files to fail as they contained the open and close quotes symbol. I changed to normal quote symbol and build goes fine. See diff below.



diff --git a/hardware/linkset/src/main/java/org/mobicents/ss7/linkset/oam/LinksetState.java b/hindex ff0d00c..10181e5 100644
--- a/hardware/linkset/src/main/java/org/mobicents/ss7/linkset/oam/LinksetState.java
+++ b/hardware/linkset/src/main/java/org/mobicents/ss7/linkset/oam/LinksetState.java
@@ -29,7 +29,7 @@ package org.mobicents.ss7.linkset.oam;
  */
 public interface LinksetState {
     /**
-     * Indicates the linkset does not have any "available" links and cannot transport traf+     * Indicates the linkset does
 not have any "available" links and cannot transport traffic
      */
     int UNAVAILABLE = 1;

diff --git a/m3ua/api/src/main/java/org/mobicents/protocols/ss7/m3ua/parameter/AffectedPointCodindex 17e5c82..1075c97 100644
--- a/m3ua/api/src/main/java/org/mobicents/protocols/ss7/m3ua/parameter/AffectedPointCode.java
+++ b/m3ua/api/src/main/java/org/mobicents/protocols/ss7/m3ua/parameter/AffectedPointCode.java
@@ -36,7 +36,7 @@ public interface AffectedPointCode extends Parameter {
      * <p>
      * To make it easier to identify multiple point codes, ranges can be used as well. The mas      * within the point code.
 For example, if the mask contains a value of 2, this would indic-     * the point code are a "wild card."
+     * the point code are a "wild card."
      * </p>
      *
      * @return returns the mask
diff --git a/map/map-api/src/main/java/org/mobicents/protocols/ss7/map/api/service/lsm/ProvideSindex 20e25aa..bddf7f1 100644
--- a/map/map-api/src/main/java/org/mobicents/protocols/ss7/map/api/service/lsm/ProvideSubscrib+++ b/map/map-api/src/main/jav
a/org/mobicents/protocols/ss7/map/api/service/lsm/ProvideSubscrib@@ -104,8 +104,8 @@ public interface ProvideSubscriberLocati
onResponse extends LsmMessage {

     /**
      * AgeOfLocationInformation ::= INTEGER (0..32767) -- the value represents the elapsed tim-     * network contact of the
 mobile station (i.e. the actuality of the -- location informatio-     * the MS is currently in contact with the -- network -
- value "32767" indicates that +     * network contact of the mobile station (i.e. the actuality of the -- location informati
o+     * the MS is currently in contact with the -- network -- value "32767" indicates that the       * -- 32767 minutes old
      *
      * @return
diff --git a/map/map-api/src/main/java/org/mobicents/protocols/ss7/map/api/service/lsm/Subscribindex e5adc58..ce08fb3 100644
--- a/map/map-api/src/main/java/org/mobicents/protocols/ss7/map/api/service/lsm/SubscriberLocat+++ b/map/map-api/src/main/jav
a/org/mobicents/protocols/ss7/map/api/service/lsm/SubscriberLocat@@ -88,8 +88,8 @@ public interface SubscriberLocationReportR
equest extends LsmMessage {

     /**
      * AgeOfLocationInformation ::= INTEGER (0..32767) -- the value represents the elapsed tim-     * network contact of the
 mobile station (i.e. the actuality of the -- location informatio-     * the MS is currently in contact with the -- network -
- value "32767" indicates that +     * network contact of the mobile station (i.e. the actuality of the -- location informati
o+     * the MS is currently in contact with the -- network -- value "32767" indicates that the       * -- 32767 minutes old
      *
      * @return

If you think it is ok I can try to do a PR with only those changes? Probably others building jss7 will face the same issue.

**Comment by vetss** [comment]
Hello @f-alonso-vendrell 

I do not fully understand what you were describing in the provious message. Feel free to create a PR if you think it fix some bugs based on master branch. I will check it.

**Comment by f-alonso-vendrell** [comment]
Just created new PR with all Aug 2nd comments. 

Please confirm that the expected behavior is:
CLIENT     SERVER
--- SRI Req -->
<-- SRI Res ---

<-- SLR Req --
--- SLR Res -->

No remove unneeded methods.

**Comment by FerUy** [comment]
Hi @f-alonso-vendrell ... the expected behaviour is this one (assuming entities interacting in core network with the GMLC are HLR, MSC/VLR):

CLIENT ----------------------  SERVER

GMLC -----SRIforLCS Req---->HLR
GMLC<----SRIforLCS Res----- HLR

GMLC --------PSL Req------->MSC/VLR
GMLC<-------PSL Res-------- MSC/VLR

(After event detected in RAN and consequently in MSC/VLR)

GMLC <-------  SLR     ------- MSC/VLR
GMLC ------- SLR Ack ------->MSC/VLR

So, your SRI and SLR expected behaviour are correct (I don't see PSL though, which is how you trigger a SLR from the RAN and forwarded by the MSC/VLR... maybe you didn't put that on purpose... ?)

More detailed in attached image, which is an enhancement of the previous one above, including SLR signals after one event detection -could be more- (not all parameters included for the sake of simplicity and clarity of the diagram -otherwise it would be busier than it already is).
![gmlc_signal_flow_umts](https://cloud.githubusercontent.com/assets/14883482/18205967/19d2d224-70fb-11e6-9620-9fd65f8e1dee.png)

**Comment by f-alonso-vendrell** [comment]
Hello @FerUy . Thanks for the clarification. Wanted to make sure that we were on right track. 

We jumped over PSL basically due to lack of knowledge when we divided work. So currently PSL (Request/Response) is not implemented yet in the simulator and is the next interaction that we will do.

It would be very helpful if you could tell us which of the optional parameters in the request response should be included, just like we did on the SLR.

Again, thanks!

**Comment by FerUy** [comment]
Hi @f-alonso-vendrell, thanks for the clarification. Indeed PSL implementation is a must!

Please find attached the parameter definitions extracted from 3GPP MAP specification for either operations. I put in bold those who will be needed. Some of them are mandatory (M), some conditional (C), some service-user option (U) or service-provider option (O).

Eventually we'll need to implement all of them, but on a first stage, the ones in bold will be the ones. Please feel free to ask if anything is not clear.

Best regards
[3GPP-29.002_MAP_PSL-LSR.docx](https://github.com/RestComm/jss7/files/452669/3GPP-29.002_MAP_PSL-LSR.docx)

**Comment by vetss** [comment]
Hello @f-alonso-vendrell 

1) I have committed your updates from https://github.com/RestComm/jss7/pull/167 into master branch. Please synch to it before your furter work.
Thank you for your gerate work.

2) I have tried to test Simulator but ti fails to load config data from xml file so config stroring is impossible now. Can you please check / fix it so we can continue of code testing

**Comment by f-alonso-vendrell** [comment]
Hello @vetss 

   I will sync, and see the data load fails, that happened to me once, but thought it was ok now. In case you want to contact me because of an issue you can do so at falonso at csc dot com

**Comment by vetss** [comment]
This is an error when starting of SS7 Simulator (failure to load config data from xml file).
I am also attaching the config file that is created by SS7 Simulator and gives an error when loading.

Please check this case.

[s_simulator2.zip](https://github.com/RestComm/jss7/files/510986/s_simulator2.zip)

java.lang.IllegalStateException: Not a start element (END_ELEMENT)
	at javolution.xml.stream.XMLStreamReaderImpl.illegalState(XMLStreamReaderImpl.java:1469)
	at javolution.xml.stream.XMLStreamReaderImpl.getAttributeValue(XMLStreamReaderImpl.java:1255)
	at javolution.xml.XMLFormat$InputElement.getAttribute(XMLFormat.java:511)
	at org.mobicents.protocols.ss7.tools.simulator.tests.lcs.TestMapLcsClientConfigurationData$1.read(TestMapLcsClientConfigurationData.java:391)
	at org.mobicents.protocols.ss7.tools.simulator.tests.lcs.TestMapLcsClientConfigurationData$1.read(TestMapLcsClientConfigurationData.java:1)
	at javolution.xml.XMLFormat$InputElement.readInstanceOf(XMLFormat.java:467)
	at javolution.xml.XMLFormat$InputElement.get(XMLFormat.java:410)
	at org.mobicents.protocols.ss7.tools.simulator.common.ConfigurationData$1.read(ConfigurationData.java:350)
	at org.mobicents.protocols.ss7.tools.simulator.common.ConfigurationData$1.read(ConfigurationData.java:1)
	at javolution.xml.XMLFormat$InputElement.readInstanceOf(XMLFormat.java:467)
	at javolution.xml.XMLFormat$InputElement.get(XMLFormat.java:410)
	at javolution.xml.XMLObjectReader.read(XMLObjectReader.java:277)
	at org.mobicents.protocols.ss7.tools.simulator.management.TesterHost.load(TesterHost.java:868)
	at org.mobicents.protocols.ss7.tools.simulator.management.TesterHost.<init>(TesterHost.java:220)
	at org.mobicents.protocols.ss7.tools.simulatorgui.ConnectionForm.startLocal(ConnectionForm.java:176)
	at org.mobicents.protocols.ss7.tools.simulatorgui.ConnectionForm.access$3(ConnectionForm.java:170)
	at org.mobicents.protocols.ss7.tools.simulatorgui.ConnectionForm$2.actionPerformed(ConnectionForm.java:147)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2018)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2341)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at java.awt.Component.processMouseEvent(Component.java:6516)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3321)
	at java.awt.Component.processEvent(Component.java:6281)
	at java.awt.Container.processEvent(Container.java:2229)
	at java.awt.Component.dispatchEventImpl(Component.java:4872)
	at java.awt.Container.dispatchEventImpl(Container.java:2287)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4832)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4492)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4422)
	at java.awt.Container.dispatchEventImpl(Container.java:2273)
	at java.awt.Window.dispatchEventImpl(Window.java:2719)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:747)
	at java.awt.EventQueue.access$300(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:706)
	at java.awt.EventQueue$3.run(EventQueue.java:704)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.awt.EventQueue$4.run(EventQueue.java:718)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:717)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)

**Comment by f-alonso-vendrell** [comment]
Hello @vetss 
   Managed to solve this, one question though. To solve this I had to move one xml element (Integer) to attribute. Is there a caveat for javolution to do so? I've seen that all non-String data is being set as attribute.
   The good news is that now it loads fine, so I will be doing a new PR that solves this.

**Comment by vetss** [comment]
@f-alonso-vendrell 

as for xml encoding:
- where to put parameters (attribute / element) is your choice. Generally just look how your serialized xml file looks like. It must be good readable. Therefore we put usually short values into attributes but long String into elements.
- "I've seen that all non-String data is being set as attribute." - this is not mandatory but reasonable in the most cases.
- when you add new elements you need to think of backup compatibilities. This mean that an xml file from old versions must be accepted. If you change a set / a format of a paramter, add a new parameter (not change a meaning of an old parameter) and removew an old one.
- do not add a new parameter at the of a parameters list, add parameters somewhere in the middle of a list (else we will have reading Exception)

**Comment by vetss** [comment]
@f-alonso-vendrell 

I have added your update https://github.com/RestComm/jss7/pull/169
I have not made moer tests. I will try to find time for tests in short time.

**Comment by f-alonso-vendrell** [comment]
Hello @FerUy I am coding geranPositioningData, utranPositioningData for PSL Response, and in order to build it it requires a byte[], for what I have seen it requires a minimum of 2 and a max of 10. How should that be presented in the Params GUI? As a String with only HexValues e.g. "F3F2F1"?

**Comment by FerUy** [comment]
Hi @f-alonso-vendrell. I faced the same issue for LTE positioning (where Diameter is used). 
I believe a String with Hex values is OK (unfortunately we don't have a trace of that yet).

**Comment by vetss** [comment]
@FerUy - is this issue fixed ?
Can we close it or we need to postpone to the next release (8.1) ?

**Comment by FerUy** [comment]
No @vetss, let's postpone it for next release 8.1.

**Comment by FerUy** [comment]
Closing. This work is being completed on commercial version of jSS7.
